secure_message/common/utilities.py

```diff
@@ -70,56 +70,76 @@ def process_paginated_list(paginated_list, host_url, user, message_args, endpoin
+def get_from_details(messages):
+    """looks up the details for the from users"""
+    update_internal_messages_from(messages)
+    from_details = party.get_user_details(get_messages_from_external(messages))
+    external_messages = [x for x in messages if x['from_internal'] is False]
+
+    for message in messages:
+        if message in external_messages:
+            message.update({'@msg_from': from_details[0]})
 
-    details = []
+    return messages
 
-    for ru in rus:
 
-        detail = party.get_business_details(ru)
+def get_messages_from_external(messages):
+    external_msgs = []
+    uuid_from = []
 
-        if detail is not None:
-            details.append(detail)
-        else:
-            logger.info('No details found for RU ID', ru=ru)
+    len_list = len(messages)
 
-    return details
+    update_internal_messages_from(messages)
 
+    for i in range(0, len_list):
+        if not messages[i]['from_internal']:
+            external_msgs.append(messages[i])
+    for uuid in external_msgs:
+        if uuid["msg_from"] not in uuid_from:
+            uuid_from.append(uuid["msg_from"])
+    return uuid_from
 
-def add_to_and_from_details(messages):
-    """Adds user details for sender and recipient"""
+
+def update_internal_messages_from(messages):
     for message in messages:
-        message.update({'@msg_from': _get_from_details(message), '@msg_to': _get_to_details(message)})
-    return messages
+        if message["from_internal"]:
+            message.update({"@msg_from": internal_user_service.get_user_details(message["msg_from"])})
 
 
-def _get_from_details(message):
-    """looks up the details for the from users"""
-    logger.info("Getting user from details", is_internal=message['from_internal'], user_uuid=message['msg_from'])
-    if message['from_internal']:
-        from_details = internal_user_service.get_user_details(message['msg_from'])
-    else:
-        from_details = party.get_user_details(message['msg_from'])
-    return from_details
+def get_to_details(messages):
+    """looks up the details for the to users"""
 
+    update_external_messages_to(messages)
 
-def _get_to_details(message):
-    """looks up the details of all the to users"""
-    user_details = []
-    if message['from_internal']:
-        to_user_details_finder = party.get_user_details
-    else:
-        to_user_details_finder = internal_user_service.get_user_details
+    to_details = party.get_user_details(get_internal_messages_to_uuid(messages))
+    internal_messages = [x for x in messages if x['from_internal'] is True]
+
+    for msg in messages:
+        if msg in internal_messages:
+            msg.update({'@msg_to': to_details})
+
+    return messages
+
+
+def update_external_messages_to(messages):
+    for message in messages:
+        if not message["from_internal"]:
+            message.update({"@msg_to": internal_user_service.get_user_details(message["msg_to"][0])})
```

Comment:
Not sure if I'm misunderstanding this.  But you get all the external messages, then you repeatedly call the internal_user_service for each one?